### PR TITLE
Publish information about Infinispan availability in lb-check if MULTI_SITE is enabled

### LINKED ISSUE #25077: Publish information about Infinispan availability in lb-check if MULTI_SITE is enabled
### Description

Introduce a new check available for the load balancer to decide whether this site should receive load or not.

This is Keycloak implementation part of the following issue for Keycloak benchmark:
https://github.com/keycloak/keycloak-benchmark/issues/591

### PR DESCRIPTION
**Pull request by mhajas**
Closes #25077 

We didn't use the health endpoints, as those would hopefully soon move to another port and shouldn't be available from the outside. It is the first time we're using a non-blocking RESTEasy request.

### CODE REVIEW

File: common/src/main/java/org/keycloak/common/Profile.java
Comment (pruivo) [inline common/src/main/java/org/keycloak/common/Profile.java:101]: I'm not sure about the name. The load balancer checker is useful outside of cross-dc support; for example, if a user configured the embedded Infinispan with a database (JDBC cache store). Probably some name related to persistent sessions would be a better fit?

**Review by pruivo — COMMENTED** [review]
LGTM, just a minor comment about the feature name.

**Comment by mhajas** [comment]
The failed CI does not seem to be caused by my changes. I saw something similar happened here: https://github.com/keycloak/keycloak/actions/runs/7016449133

File: common/src/main/java/org/keycloak/common/Profile.java
Comment (ahus1) [reply common/src/main/java/org/keycloak/common/Profile.java]: I agree this check `lb-check` has the potential to be useful in other places as well. For now, I'd like to narrow its use down to multi-site as this gives us the chance to modify it as needed which the check we want to include. I was even considering to name the URL ` lb-check-for-multi-site`, but let's be a little bit ambitious here that it might be useful in other scenarios as well  :-)

**Review by ahus1 — COMMENTED** [review]
@mhajas - thank you for this PR, it looks good to me. I added a bunch of JavaDoc to explain more about the why and the context in which those checks are being executed. If that looks good to you, I'll approve and merge it (once all builds are green, which seems to be a problem today due to some other services failing).

**Comment by ahus1** [comment]
@pruivo / @mhajas: I tried to stick to "multi-site" consistently, as this is what is now also in the docs. That's also why I don't use "cross-dc" so much, as that doesn't contain "site" which is the term used consistently inside of Infinispan, and which is now also used in Keycloak's docs.

**Review by mhajas — COMMENTED** [review]
Thank you for the addition @ahus1. It seems I am not able to approve this PR since I am the author but I agree with your changes. I think we should wait for @keycloak/cloud-native team to review.

**Review by ahus1 — DISMISSED** [review]

**Review by mabartos — COMMENTED** [review]
@mhajas @ahus1 The PR in general LGTM. I've just added some nitpicks and one suggestion/question (if we have some time to change it until the CI is up again). Not a blockers. I'd also suggest adding a simple test to verify the endpoint really works. 

@shawkins It'd be good to have your eyes on this to consider a real scenario with LB.

File: model/infinispan/src/main/java/org/keycloak/connections/infinispan/InfinispanMultiSiteLoadBalancerCheckProviderFactory.java
Comment (mabartos) [reply model/infinispan/src/main/java/org/keycloak/connections/infinispan/InfinispanMultiSiteLoadBalancerCheckProviderFactory.java:45]: But I am not sure if it is feasible in such form as it'd be required to override the `getCache(String, boolean)` method in custom implementations. Otherwise, the caches would be created in the `InfinispanMultiSiteLoadBalancerCheckProvider`, which is not what we want :) Or don't make the method default. 

In the end, it's not probably a crucial point to support it in such a way.

**Comment by mhajas** [comment]
Thanks for the review @mabartos, I addressed your comments. Could you please review again?

**Review by mabartos — APPROVED** [review]
@mhajas LGTM, thanks.

**Review by ahus1 — APPROVED** [review]
Tested this manually locally and reviewed the code, approving. @mhajas - please create a separate issue to create a test case similar for HealthDistTest to test at least the happy path.

**Comment by mhajas** [comment]
> Tested this manually locally and reviewed the code, approving. @mhajas - please create a separate issue to create a test case similar for HealthDistTest to test at least the happy path.

Done: https://github.com/keycloak/keycloak/issues/25113

**Pull request merged**
